File: platform/analysis-api/src/com/intellij/codeHighlighting/HighlightDisplayLevel.java
// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.codeHighlighting;

import com.intellij.icons.AllIcons;
import com.intellij.lang.annotation.HighlightSeverity;
import com.intellij.openapi.editor.colors.CodeInsightColors;
import com.intellij.openapi.editor.colors.EditorColorsManager;
import com.intellij.openapi.editor.colors.TextAttributesKey;
import com.intellij.openapi.editor.markup.TextAttributes;
import com.intellij.openapi.util.Comparing;
import com.intellij.openapi.util.Pair;
import com.intellij.openapi.util.text.StringUtil;
import com.intellij.ui.JBColor;
import com.intellij.ui.scale.JBUIScale;
import com.intellij.util.ObjectUtils;
import com.intellij.util.ui.ColorIcon;
import com.intellij.util.ui.ColorizeProxyIcon;
import com.intellij.util.ui.EmptyIcon;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.*;
import java.util.HashMap;
import java.util.Map;

public class HighlightDisplayLevel {
  private static final Map<HighlightSeverity, HighlightDisplayLevel> LEVEL_MAP = new HashMap<>();

  public HighlightDisplayLevel(@NotNull HighlightSeverity severity, @NotNull Icon icon) {
    this(severity, icon, icon);
  }

  private HighlightDisplayLevel(@NotNull HighlightSeverity severity, @NotNull Icon icon, @NotNull Icon outlineIcon) {
    this(severity);
    this.icon = icon;
    this.outlineIcon = outlineIcon;
    LEVEL_MAP.put(this.severity, this);
  }

  public HighlightDisplayLevel(@NotNull HighlightSeverity severity) {
    this.severity = severity;
  }

  public static final HighlightDisplayLevel GENERIC_SERVER_ERROR_OR_WARNING =
    new HighlightDisplayLevel(HighlightSeverity.GENERIC_SERVER_ERROR_OR_WARNING,
                              createIconPair(CodeInsightColors.GENERIC_SERVER_ERROR_OR_WARNING,
                                             AllIcons.General.InspectionsWarning, AllIcons.General.InspectionsWarningEmpty));

  public static final HighlightDisplayLevel ERROR =
    new HighlightDisplayLevel(HighlightSeverity.ERROR,
                              createIconPair(CodeInsightColors.ERRORS_ATTRIBUTES,
                                             AllIcons.General.InspectionsError, AllIcons.General.InspectionsErrorEmpty));

  public static final HighlightDisplayLevel WARNING =
    new HighlightDisplayLevel(HighlightSeverity.WARNING,
                              createIconPair(CodeInsightColors.WARNINGS_ATTRIBUTES,
                                             AllIcons.General.InspectionsWarning, AllIcons.General.InspectionsWarningEmpty));

  private static final TextAttributesKey DO_NOT_SHOW_KEY = TextAttributesKey.createTextAttributesKey("DO_NOT_SHOW");
  public static final HighlightDisplayLevel DO_NOT_SHOW = new HighlightDisplayLevel(HighlightSeverity.INFORMATION, EmptyIcon.ICON_0);
  
  public static final HighlightDisplayLevel CONSIDERATION_ATTRIBUTES = new HighlightDisplayLevel(HighlightSeverity.TEXT_ATTRIBUTES, EmptyIcon.ICON_0);

  /**
   * @deprecated use {@link #WEAK_WARNING} instead
   */
  @Deprecated(forRemoval = true)
  public static final HighlightDisplayLevel INFO = new HighlightDisplayLevel(HighlightSeverity.INFO, createIconByKey(DO_NOT_SHOW_KEY));

  public static final HighlightDisplayLevel WEAK_WARNING =
    new HighlightDisplayLevel(HighlightSeverity.WEAK_WARNING,
                              createIconPair(CodeInsightColors.WEAK_WARNING_ATTRIBUTES,
                                             AllIcons.General.InspectionsWarning, AllIcons.General.InspectionsWarningEmpty));

  public static final HighlightDisplayLevel NON_SWITCHABLE_ERROR = new HighlightDisplayLevel(HighlightSeverity.ERROR) {
    @Override
    public boolean isNonSwitchable() {
      return true;
    }
  };
  public static final HighlightDisplayLevel NON_SWITCHABLE_WARNING = new HighlightDisplayLevel(HighlightSeverity.WARNING) {
    @Override
    public boolean isNonSwitchable() {
      return true;
    }
  };

  private Icon icon = EmptyIcon.ICON_16;
  private Icon outlineIcon = EmptyIcon.ICON_16;
  private final HighlightSeverity severity;

  public static @Nullable HighlightDisplayLevel find(String name) {
    if ("NON_SWITCHABLE_ERROR".equals(name)) return NON_SWITCHABLE_ERROR;
    if ("NON_SWITCHABLE_WARNING".equals(name)) return NON_SWITCHABLE_WARNING;
    for (Map.Entry<HighlightSeverity, HighlightDisplayLevel> entry : LEVEL_MAP.entrySet()) {
      HighlightSeverity severity = entry.getKey();
      HighlightDisplayLevel displayLevel = entry.getValue();
      if (Comparing.strEqual(severity.getName(), name)) {
        return displayLevel;
      }
    }
    return null;
  }

  public static HighlightDisplayLevel find(@NotNull HighlightSeverity severity) {
    return LEVEL_MAP.get(severity);
  }

  public @NonNls String toString() {
    return severity.toString();
  }

  public @NotNull @NonNls String getName() {
    return severity.getName();
  }

  public @NotNull Icon getIcon() {
    return icon;
  }

  public @NotNull Icon getOutlineIcon() {
    return outlineIcon;
  }

  public @NotNull HighlightSeverity getSeverity(){
    return severity;
  }

  public boolean isNonSwitchable() {
    return false;
  }

  public static void registerSeverity(@NotNull HighlightSeverity severity, @NotNull TextAttributesKey key, @Nullable Icon icon) {
    Pair<Icon, Icon> iconPair = icon == null ? createIconByKey(key) : new Pair<>(icon, icon);
    HighlightDisplayLevel level = LEVEL_MAP.get(severity);
    if (level == null) {
      new HighlightDisplayLevel(severity, iconPair.first, iconPair.second);
    }
    else {
      level.icon = iconPair.first;
      level.outlineIcon = iconPair.second;
    }
  }

  public static int getEmptyIconDim() {
    return JBUIScale.scale(14);
  }

  private static Pair<Icon, Icon> createIconByKey(@NotNull TextAttributesKey key) {
    if (StringUtil.containsIgnoreCase(key.getExternalName(), "error")) {
      return createIconPair(key, AllIcons.General.InspectionsError, AllIcons.General.InspectionsErrorEmpty);
    }
    else {
      return createIconPair(key, AllIcons.General.InspectionsWarning, AllIcons.General.InspectionsWarningEmpty);
    }
  }

  private static @NotNull Pair<Icon, Icon> createIconPair(@NotNull TextAttributesKey key, @NotNull Icon first, @NotNull Icon second) {
    return new Pair<>(new ColorizedIcon(key, first), new ColorizedIcon(key, second));
  }

  public static @NotNull Icon createIconByMask(final Color renderColor) {
    return new MyColorIcon(getEmptyIconDim(), renderColor);
  }

  private static final class MyColorIcon extends ColorIcon implements ColoredIcon {
    MyColorIcon(int size, @NotNull Color color) {
      super(size, color);
    }

    @Override
    public Color getColor() {
      return getIconColor();
    }
  }

  public interface ColoredIcon {
    Color getColor();
  }

  private static @Nullable Color getColorFromAttributes(@NotNull TextAttributesKey key) {
    final EditorColorsManager manager = EditorColorsManager.getInstance();
    if (manager != null) {
      TextAttributes attributes = manager.getGlobalScheme().getAttributes(key);
      Color stripe = attributes == null ? null : attributes.getErrorStripeColor();
      if (stripe != null) return stripe;
      if (attributes != null) {
        Color effectColor = attributes.getEffectColor();
        if (effectColor != null) {
          return effectColor;
        }
        Color foregroundColor = attributes.getForegroundColor();
        if (foregroundColor != null) {
          return foregroundColor;
        }
        return attributes.getBackgroundColor();
      }
      return null;
    }
    TextAttributes defaultAttributes = key.getDefaultAttributes();
    if (defaultAttributes == null) {
      defaultAttributes = TextAttributes.ERASE_MARKER;
    }
    return defaultAttributes.getErrorStripeColor();
  }

  private static final class ColorizedIcon extends ColorizeProxyIcon implements ColoredIcon {
    private final TextAttributesKey myKey;

    private ColorizedIcon(@NotNull TextAttributesKey key, @NotNull Icon baseIcon) {
      super(baseIcon);
      myKey = key;
    }

    @Override
    public @NotNull Color getColor() {
      return ObjectUtils.notNull(getColorFromAttributes(myKey), JBColor.GRAY);
    }
  }
}

File: platform/analysis-api/src/com/intellij/codeHighlighting/HighlightDisplayLevel.kt
// Copyright 2000-2023 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
@file:Suppress("ReplaceGetOrSet", "ReplacePutWithAssignment")

package com.intellij.codeHighlighting

import com.intellij.icons.AllIcons
import com.intellij.lang.annotation.HighlightSeverity
import com.intellij.openapi.editor.colors.CodeInsightColors
import com.intellij.openapi.editor.colors.EditorColorsManager
import com.intellij.openapi.editor.colors.TextAttributesKey
import com.intellij.openapi.editor.markup.TextAttributes
import com.intellij.ui.IconManager
import com.intellij.ui.JBColor
import com.intellij.ui.scale.JBUIScale
import com.intellij.util.ui.ColorIcon
import com.intellij.util.ui.EmptyIcon
import org.jetbrains.annotations.ApiStatus.Internal
import org.jetbrains.annotations.NonNls
import java.awt.Color
import java.awt.Component
import java.awt.Graphics
import java.util.concurrent.ConcurrentHashMap
import javax.swing.Icon

open class HighlightDisplayLevel(severity: HighlightSeverity) {
  @Internal
  data class SeverityDescriptor(
    @JvmField val severity: HighlightSeverity,
    @JvmField val attributesKey: TextAttributesKey,
    @JvmField val icon: Icon?,
  )

  constructor(severity: HighlightSeverity, icon: Icon) : this(severity = severity, icon = icon, outlineIcon = icon)

  private constructor(severity: HighlightSeverity, icon: Icon, outlineIcon: Icon) : this(severity) {
    this.icon = icon
    this.outlineIcon = outlineIcon
    @Suppress("LeakingThis")
    registerLevel(this)
  }

  var severity: HighlightSeverity = severity
    private set

  var icon: Icon = EmptyIcon.ICON_16
    private set
  var outlineIcon: Icon = EmptyIcon.ICON_16
    private set

  companion object {
    private data class LevelPresentation(
      @JvmField val key: TextAttributesKey,
      @JvmField val icon: Icon?,
      @JvmField val outlineIcon: Icon?,
    )

    private data class LevelStateBackup(
      @JvmField val severity: HighlightSeverity,
      @JvmField val icon: Icon,
      @JvmField val outlineIcon: Icon,
    )

    private val LEVEL_MAP = ConcurrentHashMap<String, HighlightDisplayLevel>()
    private val LEVEL_LOCK = Any()
    private val providedLevelNames = HashSet<String>()
    private val shadowedLevelBackups = HashMap<String, LevelStateBackup>()

    private fun registerLevel(level: HighlightDisplayLevel) {
      LEVEL_MAP[level.severity.name] = level
    }

    private fun createPresentation(
      key: TextAttributesKey,
      icon: Icon?,
      outlineIcon: Icon? = icon,
    ): LevelPresentation {
      return LevelPresentation(key = key, icon = icon, outlineIcon = outlineIcon)
    }

    private fun updateLevel(level: HighlightDisplayLevel, severity: HighlightSeverity, presentation: LevelPresentation) {
      val (icon, outlineIcon) = createIcons(presentation)
      level.severity = severity
      level.icon = icon
      level.outlineIcon = outlineIcon
      registerLevel(level)
    }

    private fun restoreLevel(levelName: String) {
      val level = LEVEL_MAP[levelName] ?: return
      val backup = shadowedLevelBackups.remove(levelName)
      if (backup == null) {
        LEVEL_MAP.remove(levelName)
      }
      else {
        level.severity = backup.severity
        level.icon = backup.icon
        level.outlineIcon = backup.outlineIcon
        registerLevel(level)
      }
    }

    private fun createIcons(presentation: LevelPresentation): Pair<Icon, Icon> {
      presentation.icon?.let { icon ->
        return icon to (presentation.outlineIcon ?: icon)
      }

      return if (presentation.key.externalName.contains("error", ignoreCase = true)) {
        HighlightDisplayLevelColorizedIcon(key = presentation.key, baseIcon = AllIcons.General.InspectionsError) to
        HighlightDisplayLevelColorizedIcon(key = presentation.key, baseIcon = AllIcons.General.InspectionsErrorEmpty)
      }
      else {
        HighlightDisplayLevelColorizedIcon(key = presentation.key, baseIcon = AllIcons.General.InspectionsWarning) to
        HighlightDisplayLevelColorizedIcon(key = presentation.key, baseIcon = AllIcons.General.InspectionsWarningEmpty)
      }
    }

    private fun presentationFrom(descriptor: SeverityDescriptor): LevelPresentation {
      return createPresentation(key = descriptor.attributesKey, icon = descriptor.icon)
    }

    private fun createHighlightDisplayLevel(
      severity: HighlightSeverity,
      key: TextAttributesKey,
      icon: Icon,
      outlineIcon: Icon,
    ): HighlightDisplayLevel {
      val presentation = createPresentation(key = key, icon = icon, outlineIcon = outlineIcon)
      val (effectiveIcon, effectiveOutlineIcon) = createIcons(presentation)
      return HighlightDisplayLevel(
        severity = severity,
        icon = effectiveIcon,
        outlineIcon = effectiveOutlineIcon,
      )
    }

    @JvmField
    val GENERIC_SERVER_ERROR_OR_WARNING: HighlightDisplayLevel = createHighlightDisplayLevel(
      severity = HighlightSeverity.GENERIC_SERVER_ERROR_OR_WARNING,
      key = CodeInsightColors.GENERIC_SERVER_ERROR_OR_WARNING,
      icon = AllIcons.General.InspectionsWarning,
      outlineIcon = AllIcons.General.InspectionsWarningEmpty,
    )

    @JvmField
    val ERROR: HighlightDisplayLevel = createHighlightDisplayLevel(
      severity = HighlightSeverity.ERROR,
      key = CodeInsightColors.ERRORS_ATTRIBUTES,
      icon = AllIcons.General.InspectionsError,
      outlineIcon = AllIcons.General.InspectionsErrorEmpty,
    )

    @JvmField
    val WARNING: HighlightDisplayLevel = createHighlightDisplayLevel(
      severity = HighlightSeverity.WARNING,
      key = CodeInsightColors.WARNINGS_ATTRIBUTES,
      icon = AllIcons.General.InspectionsWarning,
      outlineIcon = AllIcons.General.InspectionsWarningEmpty,
    )

    private val DO_NOT_SHOW_KEY = TextAttributesKey.createTextAttributesKey("DO_NOT_SHOW")

    @JvmField
    val DO_NOT_SHOW: HighlightDisplayLevel = HighlightDisplayLevel(severity = HighlightSeverity.INFORMATION,
                                                                   icon = EmptyIcon.ICON_0)

    @JvmField
    val CONSIDERATION_ATTRIBUTES: HighlightDisplayLevel = HighlightDisplayLevel(severity = HighlightSeverity.TEXT_ATTRIBUTES,
                                                                                icon = EmptyIcon.ICON_0)


    @Suppress("DEPRECATION")
    @JvmField
    @Deprecated("use {@link #WEAK_WARNING} instead")
    val INFO: HighlightDisplayLevel = createHighlightDisplayLevel(severity = HighlightSeverity.INFO,
                                                                  key = DO_NOT_SHOW_KEY, icon = AllIcons.General.InspectionsWarning,
                                                                  outlineIcon = AllIcons.General.InspectionsWarningEmpty)

    @JvmField
    val WEAK_WARNING: HighlightDisplayLevel = createHighlightDisplayLevel(
      severity = HighlightSeverity.WEAK_WARNING,
      key = CodeInsightColors.WEAK_WARNING_ATTRIBUTES,
      icon = AllIcons.General.InspectionsWarning,
      outlineIcon = AllIcons.General.InspectionsWarningEmpty,
    )

    @JvmField
    val NON_SWITCHABLE_ERROR: HighlightDisplayLevel = object : HighlightDisplayLevel(HighlightSeverity.ERROR) {
      override val isNonSwitchable: Boolean
        get() = true
    }

    @JvmField
    val NON_SWITCHABLE_WARNING: HighlightDisplayLevel = object : HighlightDisplayLevel(HighlightSeverity.WARNING) {
      override val isNonSwitchable: Boolean
        get() = true
    }

    @JvmStatic
    fun find(name: String?): HighlightDisplayLevel? {
      return when (name) {
        "NON_SWITCHABLE_ERROR" -> NON_SWITCHABLE_ERROR
        "NON_SWITCHABLE_WARNING" -> NON_SWITCHABLE_WARNING
        else -> name?.let { LEVEL_MAP[it] }
      }
    }

    @JvmStatic
    fun find(severity: HighlightSeverity): HighlightDisplayLevel? = LEVEL_MAP[severity.name]

    @JvmStatic
    fun registerSeverity(severity: HighlightSeverity, key: TextAttributesKey, icon: Icon?) {
      val presentation = createPresentation(key = key, icon = icon)
      synchronized(LEVEL_LOCK) {
        val level = LEVEL_MAP[severity.name]
        if (level == null) {
          val (effectiveIcon, outlineIcon) = createIcons(presentation)
          HighlightDisplayLevel(severity = severity, icon = effectiveIcon, outlineIcon = outlineIcon)
        }
        else {
          updateLevel(level, severity, presentation)
        }
      }
    }

    @JvmStatic
    fun syncProvidedSeverities(severities: Map<String, SeverityDescriptor>) {
      synchronized(LEVEL_LOCK) {
        val remainingProvided = LinkedHashMap<String, HighlightDisplayLevel>()
        for (providedLevelName in providedLevelNames) {
          LEVEL_MAP[providedLevelName]?.let { remainingProvided[providedLevelName] = it }
        }

        for ((levelName, descriptor) in severities) {
          val level = LEVEL_MAP[levelName]
          if (level != null && levelName !in providedLevelNames) {
            shadowedLevelBackups.putIfAbsent(levelName, LevelStateBackup(level.severity, level.icon, level.outlineIcon))
          }

          val presentation = presentationFrom(descriptor)
          val severity = descriptor.severity
          if (level == null) {
            val (icon, outlineIcon) = createIcons(presentation)
            HighlightDisplayLevel(severity = severity, icon = icon, outlineIcon = outlineIcon)
          }
          else {
            updateLevel(level, severity, presentation)
          }

          providedLevelNames.add(levelName)
          remainingProvided.remove(levelName)
        }

        for (removedLevelName in remainingProvided.keys) {
          providedLevelNames.remove(removedLevelName)
          restoreLevel(removedLevelName)
        }
      }
    }

    @JvmStatic
    val emptyIconDim: Int
      get() = JBUIScale.scale(14)

    @JvmStatic
    fun createIconByMask(renderColor: Color): Icon {
      return HighlightDisplayLevelColorIcon(size = emptyIconDim, color = renderColor)
    }
  }

  override fun toString(): @NonNls String = severity.toString()

  val name: @NonNls String
    get() = severity.name

  open val isNonSwitchable: Boolean
    get() = false
}

sealed interface HighlightDisplayLevelColoredIcon : Icon {
  fun getColor(): Color
  fun getIcon(): Icon
}

private class HighlightDisplayLevelColorIcon(size: Int, color: Color) : ColorIcon(size, color), HighlightDisplayLevelColoredIcon {
  override fun getColor(): Color = iconColor
  override fun getIcon(): Icon = this
}

private class HighlightDisplayLevelColorizedIcon(private val key: TextAttributesKey,
                                                 baseIcon: Icon) : Icon, HighlightDisplayLevelColoredIcon {
  private val baseIcon = IconManager.getInstance().colorizedIcon(baseIcon = baseIcon, colorProvider = ::getColor)

  private var lastEditorColorManagerModCounter = -1L
  private var lastColor: Color? = null

  override fun getColor(): Color = getColorFromAttributes(key) ?: JBColor.GRAY
  override fun getIcon(): Icon  = baseIcon

  private fun getColorFromAttributes(key: TextAttributesKey): Color? {
    val editorColorManager = EditorColorsManager.getInstance()
                             ?: return (key.getDefaultAttributes() ?: TextAttributes.ERASE_MARKER).errorStripeColor
    lastColor?.takeIf { editorColorManager.schemeModificationCounter == lastEditorColorManagerModCounter}?.let {
      return it
    }

    val attributes = editorColorManager.getGlobalScheme().getAttributes(key)
    val stripe = attributes?.errorStripeColor
    val result = when {
      stripe != null -> stripe
      attributes == null -> null
      else -> attributes.effectColor ?: attributes.foregroundColor ?: attributes.backgroundColor
    }

    lastEditorColorManagerModCounter = editorColorManager.schemeModificationCounter
    lastColor = result
    return result
  }

  override fun paintIcon(c: Component?, g: Graphics?, x: Int, y: Int) {
    baseIcon.paintIcon(c, g, x, y)
  }

  override fun getIconWidth(): Int = baseIcon.iconWidth

  override fun getIconHeight(): Int = baseIcon.iconHeight
}
